### PR TITLE
[3.10] Backport the CssIdentifierRule to 3.10

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue https://github.com/joomla/joomla-cms/pull/29108 

### Summary of Changes

This here just backports the new CssIdentifierRule to 3.10 so extensions can use the new validation rule in 3.10 too.

### Testing Instructions

code review

### Expected result

Extension devs can use the new validation rule in 3.10 and 4.0

### Documentation Changes Required

This behavior will be documented together with the 4.0 PR.

### CODE REVIEW

File: libraries/src/Form/Rule/CssIdentifierRule.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	 * @return  boolean  True if the value is valid, false otherwise.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function test(\SimpleXMLElement $element, $value, $group = null, Registry $input = null, Form $form = null)
+	{
+		$value = trim($value);
+
+		// If the field is empty and not required, the field is valid.
+		$required = ((string) $element['required'] == 'true' || (string) $element['required'] == 'required');
```
Comment (SharkyKZ) [inline libraries/src/Form/Rule/CssIdentifierRule.php]:
```suggestion
		$required = ((string) $element['required'] === 'true' || (string) $element['required'] === 'required');
```

File: libraries/src/Form/Rule/CssIdentifierRule.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+		if (!$required && empty($value))
+		{
+			return true;
+		}
+
+		// Make sure we allow multible classes to be added
```
Comment (SharkyKZ) [inline libraries/src/Form/Rule/CssIdentifierRule.php]:
```suggestion
		// Make sure we allow multiple classes to be added
```

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on 497db9f673787501f3c7ceca6bc802ea3eaa6a41<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29109">issues.joomla.org/tracker/joomla-cms/29109</a>.</sub>

**Comment by mbabker** [comment]
This is B/C breaking in the context of 3.x and how the module suffix class is actually used.

**Comment by HLeithner** [comment]
> This is B/C breaking in the context of 3.x and how the module suffix class is actually used.

Why is this a B/C break? he only adds a class that's not used by the core in 3.10 and is intended for 3rd party developers to prepare there extension.

**Comment by mbabker** [comment]
Anyone who uses it cannot use it in conjunction with the module class suffix as is used in 3.x, or other class suffix structures where there is no arbitrary leading space.  The rule strips out a number of characters that would be used to indicate a modifier to the class instead of appending (i.e. with `<div class="module<?php echo $suffix; ?>">` I cannot have `$suffix = '--large'` as the rule indicates this BEM pattern is not allowed as the first characters, similar for `$suffix = '-100'` or `$suffix = '100'`).

This would be better left in 4.0 only instead of adding a class that adds some pretty hefty restrictions that would only cause end user confusion (why can I put this on X module but not on Y).

**Comment by zero-24** [comment]
> The rule strips out a number of characters that would be used to indicate a modifier to the class instead of appending (i.e. with <div class="module<?php echo $suffix; ?>"> I cannot have $suffix = '--large' as the rule indicates this BEM pattern is not allowed as the first characters, similar for $suffix = '-100' or $suffix = '100').

Just to be clear that rule is not there in stone ;) So when we have the requirement to allow --large this can be added.

Do you have some more usecases we should take a look into?

**Comment by conconnl** [comment]
Security or not, the remark in the code clearly states it will break existing sites no matter the UseCase. 

Remark: “ // Identifiers cannot start with a digit, two hyphens, or a hyphen followed by a digit.”

Any site which upgrades to 3.10 and any change done by a user in a location where a class is used, which get denied by this check will break stuff. 

This is not only irritating for the users, also for any site maintainer, developer whatsoever. 
Not to mention the negative marketing it will generate with a possible (most likely) B/C break like this. 

I’m not planning to fix sites where by any chance the class starts with -Digit, — or just digit.

**Comment by SharkyKZ** [comment]
@conconnl this introduces the validation rule but does not apply it to any existing fields in core. So it's completely optional in 3.10. The PR for 4.0 does apply the rule to all class fields.

But I agree, the regex needs to be changed to allow existing suffix uses. This applies to 4.0 as well. Even though core layouts no longer allows class suffixes, nothing should stop users and template developers from fixing this themselves.

**Comment by zero-24** [comment]
> Remark: “ // Identifiers cannot start with a digit, two hyphens, or a hyphen followed by a digit.”

Yes when you check where that code is based on you notice that they use is different that we do in our code.

So please let me know how you use this field and we make sure the validation rule supports that but we have to start somewhere.

>This is not only irritating for the users, also for any site maintainer, developer whatsoever.
Not to mention the negative marketing it will generate with a possible (most likely) B/C break like this.

As mention above there is zero impact on 3.10 installations. ;)

**Comment by zero-24** [comment]
> But I agree, the regex needs to be changed to allow existing suffix uses. This applies to 4.0 as well. Even though core layouts no longer allows class suffixes, nothing should stop users and template developers from fixing this themselves.

Please add examples of what you use so we can make sure it passes the validation rule.

File: libraries/src/Form/Rule/CssIdentifierRule.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	{
+		$value = trim($value);
+
+		// If the field is empty and not required, the field is valid.
+		$required = ((string) $element['required'] === 'true' || (string) $element['required'] === 'required');
+
+		if (!$required && empty($value))
+		{
+			return true;
+		}
```
Comment (mvanvu) [inline libraries/src/Form/Rule/CssIdentifierRule.php]: $value == 0 | false is also empty. How about this case?

Comment (zero-24) [reply libraries/src/Form/Rule/CssIdentifierRule.php]: Hi, do you mean when the string in that field is set to `0` or `false`? What happen in that case right now?

Comment (mvanvu) [reply libraries/src/Form/Rule/CssIdentifierRule.php]: Hi, if the value is set to 0 or false so the value was present. I think in this case we should continue to validate, don't return it to true.

Or maybe: `if (!$required && empty($value) && $value !== '0')` => just for eg, not a solution.

Comment (zero-24) [reply libraries/src/Form/Rule/CssIdentifierRule.php]: Hmm will have a closer look but both values would pass the validation anyway right?

Comment (SharkyKZ) [reply libraries/src/Form/Rule/CssIdentifierRule.php]: No, because:

>Identifiers cannot start with a digit ...

Comment (mvanvu) [reply libraries/src/Form/Rule/CssIdentifierRule.php]: In this case, if not required and the value is set to 0 it will return TRUE => not correct.

Comment (zero-24) [reply libraries/src/Form/Rule/CssIdentifierRule.php]: Ok got your point will take care of that. Thanks!

**Comment by zero-24** [comment]
@mbabker @conconnl @mvanvu The latest changes form the 4.x PR has been merged in here too. That should fix the issues mention here. Feel free to let me know anything that is still a problem so we can adress them.

File: libraries/src/Form/Rule/CssIdentifierRule.php (hunk 3 above)
Comment (zero-24) [reply libraries/src/Form/Rule/CssIdentifierRule.php]: I have taken care of that in the last commit thanks.

File: libraries/src/Form/Rule/CssIdentifierRule.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+		// Make sure we allow multiple classes to be added
+		$cssIdentifiers = explode(' ', $value);
+
+		foreach ($cssIdentifiers as $i => $identifier)
+		{
+			/**
+			 * The folowing regex rules are based on the Html::cleanCssIdentifier method from Drupal
```
Comment (Quy) [inline libraries/src/Form/Rule/CssIdentifierRule.php]:
```suggestion
			 * The following regex rules are based on the Html::cleanCssIdentifier method from Drupal
```

File: libraries/src/Form/Rule/CssIdentifierRule.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+			 * - 0-9 (U+0061 - U+007A)
+			 * - ISO 10646 characters U+00A1 and higher
+			 */
+			if (preg_match('/[^\\x{002D}\\x{0030}-\\x{0039}\\x{0041}-\\x{005A}\\x{005F}\\x{003A}\\x{0061}-\\x{007A}\\x{00A1}-\\x{FFFF}]/u', $identifier))
+			{
+				return false;
+			}
+
+			/**
+			 * Apply this rules only to full class identiers and not to the suffix ($i === 0).
```
Comment (Quy) [inline libraries/src/Form/Rule/CssIdentifierRule.php]:
```suggestion
			 * Apply this rules only to full class identifiers and not to the suffix ($i === 0).
```

**Comment by zero-24** [comment]
I'm going to close here as between 3.x and 4.x there are fundamentally changes in 4.x there is no sfx any more but in 3.x we usually have a sfx so better not port something back that can not be used (as in 3.x you use sfx) and would be different once updated to 4.x

**Pull request closed without merging**